runtimevar/drivertest/drivertest.go

```diff
@@ -48,5 +52,439 @@ type HarnessMaker func(t *testing.T) (Harness, error)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// The variable is decoded to a []*Message and matches the expected content.
+	if gotSlice, ok := got.Value.([]*Message); !ok {
+		t.Fatalf("got value of type %v expected []*Message", reflect.TypeOf(got.Value))
+	} else if !cmp.Equal(gotSlice, want) {
+		t.Errorf("got %v want %v", gotSlice, want)
+	}
+}
+
+func testInvalidJSON(t *testing.T, newHarness HarnessMaker) {
+	const (
+		name    = "test-config-variable"
+		content = "not-json"
+	)
+
+	h, err := newHarness(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	ctx := context.Background()
+
+	if err := h.CreateVariable(ctx, name, []byte(content)); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := h.DeleteVariable(ctx, name); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	var jsonData []*Message
+	v, err := h.MakeVar(ctx, name, runtimevar.NewDecoder(jsonData, runtimevar.JSONDecode))
+	if err != nil {
+		t.Fatal(err)
+	}
+	got, err := v.Watch(ctx)
+	if err == nil {
+		t.Errorf("got %v wanted invalid-json error", got)
+	}
+}
+
+func testUpdate(t *testing.T, newHarness HarnessMaker) {
+	const (
+		name     = "test-config-variable"
+		content1 = "hello world"
+		content2 = "goodbye world"
+	)
+
+	h, err := newHarness(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	ctx := context.Background()
+
+	// Create the variable and verify Watch sees the value.
+	if err := h.CreateVariable(ctx, name, []byte(content1)); err != nil {
+		t.Fatal(err)
+	}
+	defer func() { _ = h.DeleteVariable(ctx, name) }()
+
+	v, err := h.MakeVar(ctx, name, runtimevar.StringDecoder)
+	if err != nil {
+		t.Fatal(err)
+	}
+	got, err := v.Watch(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if got.Value.(string) != content1 {
+		t.Errorf("got %q want %q", got, content1)
+	}
+
+	// Update the variable and verify Watch sees the updated value.
+	if err := h.UpdateVariable(ctx, name, []byte(content2)); err != nil {
+		t.Fatal(err)
+	}
+	got, err = v.Watch(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if got.Value.(string) != content2 {
+		t.Errorf("got %q want %q", got, content2)
+	}
+}
+
+func testConcurrentUpdate(t *testing.T, newHarness HarnessMaker) {
+	const (
+		name     = "test-config-variable"
+		content1 = "hello world"
+		content2 = "goodbye world"
+	)
+
+	h, err := newHarness(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	ctx := context.Background()
+
+	// Create the variable and verify Watch sees the value.
+	if err := h.CreateVariable(ctx, name, []byte(content1)); err != nil {
+		t.Fatal(err)
+	}
+	defer func() { _ = h.DeleteVariable(ctx, name) }()
+
+	v, err := h.MakeVar(ctx, name, runtimevar.StringDecoder)
+	if err != nil {
+		t.Fatal(err)
+	}
+	got, err := v.Watch(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if got.Value.(string) != content1 {
+		t.Errorf("got %q want %q", got, content1)
+	}
+
+	// Update in a separate goroutine, waiting a bit to give the Watch
+	// a chance to block.
+	go func() {
+		time.Sleep(10 * time.Millisecond)
```

Comment:
There is no way to actually test for the property you are trying to test given the way the API is structured. Sleeping will not help with that. The problem is that when a driver does not meet this property, this test will become flaky and it will be tempting for someone (especially an external implementation) to dismiss this as a bad test.

There are two ways to address this:

1.  Delete the test as the property cannot be conclusively tested for.
2.  Make the failure more likely to occur by running it more times (but with no sleep). The test failure message should include a warning to the effect of "If you see this test flaking, it is indicating that there is almost certainly an issue with your driver as this is a probabilistic test."

(h/t @neild for the guidance here.)

Reply:
Done.

Reply:
I left a TODO to run the test more times when not in -record mode.